src/main/java/com/openshift/internal/restclient/ApiTypeMapper.java

```diff
@@ -0,0 +1,389 @@
+		}
+		return groups;
+	}
+	
+	private String readEndpoint(final String endpoint) {
+		try {
+			final URL url = new URL(new URL(this.baseUrl), endpoint);
+			LOGGER.debug(url.toString());
+			String response = client.get(url, IHttpConstants.DEFAULT_READ_TIMEOUT);
+			LOGGER.debug(response);
+			return response;
+		} catch (MalformedURLException | SocketTimeoutException e) {
+			LOGGER.error("Exception", e);
+			throw new OpenShiftException(e,"");
+		//HACK - This gets us around a server issue
+		} catch (HttpClientException e) {
+			if(e instanceof NotFoundException) {
+				throw new com.openshift.restclient.NotFoundException(e);
+			}
+			LOGGER.error("Unauthorized exception. Can system:anonymous get the API endpoint", e);
+			throw e;
+		}
+	}
+	
+	static class ApiGroup implements IApiGroup{
+		private final ModelNode node;
+		private final String prefix;
+		private final String path;
+		
+		ApiGroup(String prefix, ModelNode node) {
+			this.prefix = prefix;
+			this.node = node;
+			StringBuilder builder = new StringBuilder(prefix);
+			if(getName() != null) { //null name for k8e or openshift
+				builder
+				.append(FWD_SLASH)
+				.append(getName());
+			}
+			path = builder.toString();
+		}
+		
+		protected ModelNode getNode() {
+			return node;
+		}
+		
+		@Override
+		public String getPrefix() {
+			return prefix;
+		}
+
+		@Override
+		public String getName() {
+			return JBossDmrExtentions.asString(node, new HashMap<>(), "name");
+		}
+		
+		@Override
+		public Collection<String> getVersions() {
+			return JBossDmrExtentions.get(node, new HashMap<>(), "versions").asList()
+				.stream().map(n->n.get("version").asString())
+				.collect(Collectors.toList());
+		}
+		
+		@Override
+		public String getPreferedVersion() {
+			return JBossDmrExtentions.asString(node, new HashMap<>(), "preferedVersion.version");
+		}
+		
+		@Override
+		public String pathFor(String version) {
+			//add check for supported version?
+			return String.format("%s/%s", path, version);
+		}
+	}
+	
+	static class LegacyApiGroup extends ApiGroup{
+
+		LegacyApiGroup(String prefix, ModelNode node) {
+			super(prefix, node);
+		}
+		
+		@Override
+		public String getName() {
+			return null;
+		}
+
+		@Override
+		public Collection<String> getVersions() {
+			return JBossDmrExtentions.get(getNode(), new HashMap<>(), "versions").asList()
+				.stream().map(n->n.asString())
+				.collect(Collectors.toList());
+		}
+
+		@Override
+		public String getPreferedVersion() {
+			return OpenShiftAPIVersion.v1.toString();
+		}
+		
+		
+	}
+	
+	static class VersionedApiResource implements IVersionedApiResource {
+		
+		private final String prefix;
+		private final String name;
+		private final boolean namespaced;
+		private final Collection<String> capabilities = new ArrayList<>();
+		private final String version;
+		private String apiGroupName;
+		private String kind;
+		
+		VersionedApiResource(String prefix, String version, String name){
+			if(version.contains(FWD_SLASH)) {
```

Comment:
it can't be null otherwise there is an error in the response from the server.  We could add a throw or assert which would check for null.  You think this would be a better option?